File: BackEnd/Application/UseCases/Projects/GetUsersByProjectId/GetUsersByProjectIdQueryHandler.cs
using Application.DTOs.Users;
using Domain.Aggregates;
using Infrastructure.Repository;
using MediatR;
using Microsoft.EntityFrameworkCore;

namespace Application.UseCases.Projects.GetUsersByProjectId;

public class GetUsersByProjectIdQueryHandler(IRepository<Project, Guid> projectRepository)
    : IRequestHandler<GetUsersByProjectIdQuery, List<UserDto>>
{
    public async Task<List<UserDto>> Handle(GetUsersByProjectIdQuery request, CancellationToken cancellationToken)
    {
        var users = await projectRepository.Where(p => p.Id.Equals(request.ProjectId)).SelectMany(p => p.Users).Select(u => new UserDto
        {
            Id = u.Id,
            FullName = u.FullName,
            Email = u.Email,
            Status = u.Status,
        }).ToListAsync(cancellationToken);

        return users;
    }
}

File: flutter/lib/application/use_cases/projects/get_users/get_users_by_project_id_query_handler.dart
import 'package:enterprise_management/application/use_cases/projects/get_users/get_users_by_project_id_query.dart';
import 'package:enterprise_management/domain/aggregates/user/user.dart';
import 'package:enterprise_management/infrastructure/repositories/project_repository.dart';
import 'package:enterprise_management/shared_kernel/cqrs/cqrs.dart';
import 'package:enterprise_management/shared_kernel/result/result.dart';

class GetUsersByProjectIdQueryHandler extends IQueryHandler<GetUsersByProjectIdQuery, List<User>> {
  const GetUsersByProjectIdQueryHandler({required this._projectRepository});
  final IProjectRepository _projectRepository;

  @override
  Future<Result<List<User>>> handle(GetUsersByProjectIdQuery query) async {
    final res = await _projectRepository.getUsersByProjectId(projectId: query.projectId);
    return Result.success(res);
  }
}